### ISSUE TITLE
[Form] Ability to extend multiple form types with one type extension

### ISSUE THREAD

**Issue by xabbuh** [symfony/symfony#27906]
Currently, it is only possible to extend a single form type (and all its children) with a single form type extension. If you want to add an additional option to, for example, `TextType` and `ChoiceType` you will either have to create two extension classes or create an extension for the base `FormType` and thus adding the option to most other form types too.

@HeahDude and me discovered that we need to discuss if we want to support these uses cases in Form component before we can make a final decision on the implementation in #24530. Otherwise, we risk to deprecate the `getExtendedType()` method of the `FormTypeExtensionInterface` in favour of a solution which would then in turn be deprecated again in one of the next minor releases just to support the extension of multiple form types. That wouldn't be good DX-wise.

**Comment by yceruto** [comment]
I personally have never needed it, but I guess there could be a good reason to do it. 

Btw, there is another alternative to achieve it  https://github.com/symfony/symfony/issues/9507#issuecomment-112443691:
> So, since there is a solution that works right now (abstract type + subclasses) - even thought it's not the nicest one - I don't think this issue is critical.

**Comment by HeahDude** [comment]
The issue is not critical yes, means it is not worth working on a bc layer for itself.

IMHO the use case is legit, and modifying the interface already to handle auto configuration is a good opportunity to support it.

Abstract class and trait can be used currently yes, but I never saw that in action either. However I did see some `$builder->getType()->getInnerType()` or `$form->getConfig()->getType()->getInnerType()` in extensions. Even if that works, doing so adds a lot of overhead on unrelated forms/fields.

Anyway, having a clean and self explicit way to do it will prevent some trouble figuring out how this can be achieved and how ugly this can get.

**Comment by javiereguiluz** [comment]
In order to be able to evaluate this proposal better, could you please share with us some real-world examples where extending multiple form types at the same time would be needed? Thanks!

**Comment by javiereguiluz** [comment]
@HeahDude I'm afraid your example is still a theoretical scenario. We need real-world examples where this feature would have been nice to have. I don't use the Form component, but I know that you and others use it a lot ... so, have you ever wished this feature existed to solve a real problem? Thanks!

**Comment by vudaltsov** [comment]
When I made #25582, I introduced a new `input=datetime_immutable` option for 3 date/time types.
I also made a polyfill for this (https://github.com/ruvents/ruwork-polyfill-form-dti) where I had to write 1 abstract + 3 concrete extensions instead of 1 concrete. Looks like a real example!

**Comment by vudaltsov** [comment]
More than that now with Symfony 4.1 I am always using `DateTimeImmutable` in entities and to have `input=datetime_immutable` by default I also have a 1+3 workaround with `$resolver->setDefault('input', 'datetime_immutable')`.

**Comment by vudaltsov** [comment]
Sometimes both `FormType` and `ButtonType` might be extended. In core we have `Symfony\Component\Form\Extension\Validator\Type\BaseValidatorExtension` -> `FormTypeValidatorExtension`, `SubmitTypeValidatorExtension` which almost represents that situation (but not an ideal example because it's changed in one of subclasses).
